cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/DefaultPluginContext.java

```diff
@@ -99,8 +99,19 @@ public PluginProperties getPluginProperties(String pluginId) {
 
   @Override
   public <T> T newPluginInstance(String pluginId, MacroEvaluator evaluator) throws InstantiationException {
-    // todo : use pluginInstantiator#newPluginInstance(String pluginId, MacroEvaluator evaluator)
-    return newPluginInstance(pluginId);
+    try {
+      Plugin plugin = getPlugin(pluginId);
+      if (pluginInstantiator == null) {
+        throw new UnsupportedOperationException("Plugin is not supported");
+      }
+      return pluginInstantiator.newInstance(plugin, evaluator);
+    } catch (ClassNotFoundException e) {
+      // Shouldn't happen, unless there is bug in file localization
+      throw new IllegalArgumentException("Plugin class not found", e);
+    } catch (IOException e) {
+      // This is fatal, since jar cannot be expanded.
+      throw Throwables.propagate(e);
```

Comment:
InstantiationException doesn't allow you to pass an exception into its constructor. I'm not sure why the method declares that it throws and InstantiationException, I copied the implementation from the original non-macro method: 
`public <T> T newPluginInstance(String pluginId);`